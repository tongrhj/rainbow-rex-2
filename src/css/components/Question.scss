@import '../variables';
@import url('https://fonts.googleapis.com/css?family=Rancho:400');

.card {
  position: relative;
  opacity: 0.9;
  border-radius: 10px;
  backdrop-filter: blur(10px);
  background-color: $question-background;
  box-shadow: 0 10px 15px 0 $question-shadow;
  margin-bottom: 15px;
  min-width: 200px;
  min-height: 150px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  overflow: hidden;
  color: $question-text;
}

.highlight {
  color: $question-highlight;
  position: absolute;
  left: 20px;
  text-transform: uppercase;
}

.highlightAlt {
  color: $question-highlight-alt;
  position: absolute;
  right: 20px;
  text-transform: uppercase;
}

.question {
  position: relative;
  font-family: 'Rancho', serif;
  font-size: 4rem;
  margin: 0;
}

.red {
  color: $red;
}

.green {
  color: $green;
}

.blue {
  color: $blue;
}

.yellow {
  color: $yellow;
}

.purple {
  color: $purple;
}

.pink {
  color: $pink;
}

.orange {
  color: $orange;
}
