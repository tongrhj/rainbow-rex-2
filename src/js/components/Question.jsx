import questionCss from '../../css/components/Question.scss'
import Countdown from './Countdown'
import React, { Component } from 'react'

class Question extends Component {
  constructor (props) {
    super(props)

    this.state = {
      timeLeft: props.totalTime,
      addTime: props.addTime,
      addedTime: 0
    }
    this.timeout = {}
  }

  componentDidUpdate () {
    const { addTime, onTimeout } = this.props
    if (this.state.addTime !== addTime) {
      let cappedTime = Math.min(Question.defaultProps.totalTime, this.state.timeLeft + addTime)
      this.setState({
        addTime,
        timeLeft: cappedTime
      })
    }
    if (this.state.timeLeft <= 0) {
      onTimeout()
    }
  }

  componentDidMount () {
    const loop = () => {
      if (this.state.timeLeft <= 0) return
      this.setState({ timeLeft: this.state.timeLeft - 100 })
      this.timeout = setTimeout(loop, 100)
    }
    loop()
  }

  componentWillUnmount () {
    clearTimeout(this.timeout)
  }

  render () {
    const { word, colour, readColourRound, totalTime } = this.props
    return (
      <section className={questionCss.card}>
        {readColourRound
          ? (<span className={questionCss.highlight}>ðŸ“– read</span>)
          : (<span className={questionCss.highlightAlt}>see ðŸ‘€</span>)}
        <Countdown timeLeft={this.state.timeLeft} totalTime={totalTime} />
        <h2 className={`${questionCss[colour]} ${questionCss.question}`}>{word}</h2>
      </section>
    )
  }
}

Question.defaultProps = {
  totalTime: 10000,
  timeLeft: 10000,
  addTime: 0
}

export default Question
