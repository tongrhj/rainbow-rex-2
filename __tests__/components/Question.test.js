import React from 'react'
import Question from '../../src/js/components/Question'
import renderer from 'react-test-renderer'

beforeEach(() => {
  jest.useFakeTimers()
})

afterEach(() => {
  jest.clearAllTimers()
})

it('renders correctly', () => {
  const tree = renderer.create(
    <Question />
  ).toJSON()
  expect(setTimeout.mock.calls.length).toBe(1)
  expect(tree).toMatchSnapshot()
})

it('changes on time passing', () => {
  const component = renderer.create(
    <Question />
  )
  let tree = component.toJSON()
  expect(setTimeout.mock.calls.length).toBe(1)
  expect(tree).toMatchSnapshot()

  jest.runOnlyPendingTimers()
  tree = component.toJSON()
  expect(setTimeout.mock.calls.length).toBe(2)
  expect(tree).toMatchSnapshot()
})

it('changes on addTime', () => {
  const component = renderer.create(
    <Question />
  )
  let tree = component.toJSON()
  expect(setTimeout.mock.calls.length).toBe(1)
  expect(tree).toMatchSnapshot()

  jest.runOnlyPendingTimers()
  tree = component.toJSON()
  expect(setTimeout.mock.calls.length).toBe(2)
  expect(tree).toMatchSnapshot()

  const component2 = renderer.create(
    <Question addTime={1000} />
  )
  jest.runOnlyPendingTimers()
  tree = component2.toJSON()
  expect(tree).toMatchSnapshot()
})

it('changes on round type', () => {
  const component = renderer.create(
    <Question readColourRound />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  const component2 = renderer.create(
    <Question readColourRound={false} />
  )
  tree = component2.toJSON()
  expect(tree).toMatchSnapshot()
})

it('changes on colour and word', () => {
  const component = renderer.create(
    <Question colour={'red'} word={'green'} />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  const component2 = renderer.create(
    <Question colour={'green'} word={'red'} />
  )
  tree = component2.toJSON()
  expect(tree).toMatchSnapshot()
})

it('calls onTimeout after time left runs out', () => {
  const onTimeout = jest.fn()
  const component = renderer.create(
    <Question timeleft={1000} totalTime={1000} onTimeout={onTimeout} />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  jest.runAllTimers()
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  expect(onTimeout).toBeCalled()
})
