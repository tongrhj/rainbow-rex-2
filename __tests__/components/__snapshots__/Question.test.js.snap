exports[`test calls onTimeout after time left runs out 1`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "90%",
        }
      } />
  </div>
  <h2
    className="undefined question" />
</section>
`;

exports[`test calls onTimeout after time left runs out 2`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "0%",
        }
      } />
  </div>
  <h2
    className="undefined question" />
</section>
`;

exports[`test changes on addTime 1`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "99%",
        }
      } />
  </div>
  <h2
    className="undefined question" />
</section>
`;

exports[`test changes on addTime 2`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "98%",
        }
      } />
  </div>
  <h2
    className="undefined question" />
</section>
`;

exports[`test changes on addTime 3`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "98%",
        }
      } />
  </div>
  <h2
    className="undefined question" />
</section>
`;

exports[`test changes on colour and word 1`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "99%",
        }
      } />
  </div>
  <h2
    className="red question">
    green
  </h2>
</section>
`;

exports[`test changes on colour and word 2`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "99%",
        }
      } />
  </div>
  <h2
    className="green question">
    red
  </h2>
</section>
`;

exports[`test changes on round type 1`] = `
<section
  className="card">
  <span
    className="highlight">
    ğŸ“– read
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "99%",
        }
      } />
  </div>
  <h2
    className="undefined question" />
</section>
`;

exports[`test changes on round type 2`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "99%",
        }
      } />
  </div>
  <h2
    className="undefined question" />
</section>
`;

exports[`test changes on time passing 1`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "99%",
        }
      } />
  </div>
  <h2
    className="undefined question" />
</section>
`;

exports[`test changes on time passing 2`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "98%",
        }
      } />
  </div>
  <h2
    className="undefined question" />
</section>
`;

exports[`test renders correctly 1`] = `
<section
  className="card">
  <span
    className="highlightAlt">
    see ğŸ‘€
  </span>
  <div
    className="countdown">
    <div
      className="remaining"
      style={
        Object {
          "width": "99%",
        }
      } />
  </div>
  <h2
    className="undefined question" />
</section>
`;
